### PR TITLE
refactor: introduce TSM merge plans

### LINKED ISSUE #1073: refactor TSM w/ MergePlan
# Introduce `MergePlan` and refactor ALB TSM query execution

## Summary

Introduce a shared TSM query-planning model before adding support for the new PromQL outer aggregators in #1068. Replace the current classification tuple and weighted-average special case with a plan that describes fanout variants, per-variant accumulation, completeness requirements, reduction, finalization, and unsupported-query warnings.

Migrate the existing `avg -> sum + count` implementation to the generic plan executor as its first multi-variant user. Existing single-query merge behavior for deduplication, `sum`, `count`, `count_values`, `min`, `max`, and current `topk`/`bottomk` finalization must continue to work through single-variant plans.

This code is new and the package has no external consumers. No compatibility adapter or parallel legacy interface is required. Once the new path is covered and equivalent, remove `needsDualQuery`, `RewriteForWeightedAvg`, and the dedicated `serveWeightedAvg` execution path.

This issue is a prerequisite for the four operator-support sub-issues under #1068 and should merge to `main` before work on those implementations begins.

## Goals

- Give TSM one planning model for single-query and multi-query operations.
- Make fanout variant execution, contribution pairing, reduction, failure handling, metrics, capture limits, and finalization reusable by future aggregators.
- Use weighted `avg` to validate the plan abstraction without changing its successful-result semantics.
- Strengthen partial-failure correctness by preventing incomplete variant sets from contributing as though they were complete.
- Remove operator-specific dispatch flags and weighted-average fanout code.

## Universal planning requirement

Every time-series query handled by TSM must be represented by a `MergePlan`. Plans are not limited to rewritten, multi-query, or finalizable aggregations. This includes:

- queries with no outer aggregation, represented by one primary variant using deduplication;
- simple outer aggregations such as `sum`, `count`, `count_values`, `min`, and `max`, represented by one primary variant using the corresponding merge strategy;
- `group`, represented by one primary deduplication variant;
- existing rank/sort queries, represented by a rewritten primary variant plus the existing finalizer;
- weighted `avg`, represented by its sum and count variants;
- unsupported aggregations, represented by a deduplication fallback plan with an explicit inaccurate-result warning.

There must not be a parallel tuple/boolean classification path for "simple" queries. Planning is the uniform control plane for variant construction, merge strategy, completeness, reduction, finalization, warnings, response authority, and safe fast-path eligibility.

Uniform planning does not require every plan to use the full multi-variant executor. Execution may retain optimized data paths beneath the plan:

- a one-variant plan may use the existing standard fanout implementation;
- an unchanged one-variant request need not be cloned unnecessarily;
- a genuinely single-member pool may proxy directly when the plan requires no rewriting, injected-label stripping, reduction, finalization, or warning;
- a degraded multi-member pool must still use the planned response path so its incomplete-result warning is preserved.

Plan construction and validation, rather than separate request-time operator branches, must decide whether these optimizations are safe.

## Suggested planning model

The exact names and public shape are not prescribed, but the plan must express the following concepts:

```go
type MergePlan struct {
	// One request per logical query variant. Most plans have one; weighted avg
	// has sum and count variants.
	Variants []QueryVariant

	// Combines variant accumulators or per-member variant state after fanout.
	Reduction ReductionSpec

	// Optional operation applied after reduction/global accumulation.
	Finalizer FinalizerSpec

	// Determines whether a member may contribute if a required variant failed.
	Completeness CompletenessPolicy

	// Warning attached when the expression cannot be planned exactly.
	UnsupportedWarning string
}

type QueryVariant struct {
	// Stable observability name, such as "avg-sum" or "avg-count".
	Name string

	// A request rewritten by the provider and safe for concurrent fanout.
	Request *http.Request

	// Dataset strategy used to accumulate this variant's contributions.
	MergeStrategy int
}
```

An internal plan type, replacement backend-provider contract, reducer interface, or equivalent representation is acceptable. The important requirement is that the executor not switch on PromQL operator names or grow new booleans for future multi-query plans.

Every valid plan must satisfy these invariants:

- [ ] It contains at least one query variant.
- [ ] Variant names are unique and stable for observability.
- [ ] Every variant declares its merge strategy.
- [ ] Exactly one variant is designated as response authority.
- [ ] Standard reduction accepts the supported one-variant shape.
- [ ] Multi-variant reductions declare their member-completeness policy.
- [ ] An unsupported fallback plan carries a non-empty warning.
- [ ] Any single-member bypass is explicitly safe for that complete plan.

The plan should be created after the original request and query are known so the provider remains responsible for protocol-specific GET/POST rewriting and safe request cloning.

### Standard single-query plan

```text
variants:
  - primary: original or finalizer-rewritten request merge
  strategy: classified dedup/sum/min/max strategy
reduction:
  standard TSM accumulation
completeness:
  one usable primary contribution per participating member
finalizer:
  existing provider finalizer, when applicable
```

The existing standard fanout implementation may remain as an internal single-variant fast path, but it must be selected from a `MergePlan`, not from a separate classification/dispatch model.

### Weighted-average plan

```text
variants:
  - avg-sum:
      request: outer avg rewritten to sum
      merge strategy: sum
  - avg-count:
      request: outer avg rewritten to count
      merge strategy: sum
reduction:
  pair accumulated sum and count series/timestamps, then divide
completeness:
  require both variants from a member before either contribution from that member enters the global reduction
response authority:
  avg-sum variant, preserving current envelope/header behavior
finalizer:
  existing outer rank/sort finalizer, when applicable
```

### Plan construction

- [ ] Replace `ClassifyMerge(query) (strategy, needsDualQuery, warning)` with a plan-producing provider contract.
- [ ] Produce a plan for every TSM time-series request; no-aggregation, supported single-variant aggregation, and unsupported fallback queries must not bypass planning.
- [ ] Represent all existing supported merge strategies as single-variant plans where no additional query variant is required.
- [ ] Represent weighted `avg` as the two-variant plan described above.
- [ ] Move provider-specific request rewriting into plan construction or a provider-owned plan preparation step.
- [ ] Preserve the original query separately from rewritten variant queries so metadata, warnings, series pairing, and finalizers continue to use the correct logical statement.
- [ ] Ensure every variant request has an independently readable request body and no concurrently mutable request state.

### Generic execution

- [ ] Add a generic plan executor that can fan out one or more variants under the existing query concurrency and capture limits.
- [ ] Preserve deterministic member indexing for every variant.
- [ ] Collect contributions by `(variant, member)` before deciding which contributions are eligible for reduction.
- [ ] Apply the plan's completeness policy before merging contributions. For weighted `avg`, discard both contributions from a member unless both its sum and count variants are usable.
- [ ] Accumulate each variant with its configured merge strategy.
- [ ] Run the plan reducer only after all eligible variant contributions have been accumulated.
- [ ] Invoke the existing provider finalizer exactly once, after reduction.
- [ ] Keep cancellation checks around fanout, accumulation, reduction, and finalization.
- [ ] Preserve the all-fanout-failed `502` behavior and existing empty-result handling.

### Response and failure behavior

- [ ] Let the plan identify the response-authority variant used for status, headers, and response marshaling when variants differ.
- [ ] Preserve winner-header and multi-valued-header behavior from the current standard and weighted-average paths.
- [ ] Preserve partial-hit status reporting for mixed successful and failed members.
- [ ] Emit a clear warning naming the member and missing/failed variant when a completeness policy excludes that member.
- [ ] Never combine sum data from one member set with count data from a different member set and present it as a complete weighted average.
- [ ] Preserve degraded-pool warnings and the genuinely single-member fast path where it remains semantically safe.

### Observability and resource controls

- [ ] Preserve `ALBFanoutAttempts` and `ALBFanoutFailures` variant labels, including `avg-sum` and `avg-count`.
- [ ] Preserve panic recovery and failure-reason accounting for every variant and member.
- [ ] Preserve per-response and per-fanout capture-budget enforcement.
- [ ] Document whether a multi-variant plan's total capture budget is shared or applied independently per variant; do not change the current behavior accidentally during the refactor.
- [ ] Preserve configured query concurrency limits and avoid unbounded `variant count * member count` goroutine creation.

### Cleanup

- [ ] Remove the `needsDualQuery` flag and all associated branches.
- [ ] Remove `RewriteForWeightedAvg` from the provider contract and Prometheus backend.
- [ ] Remove or generalize `serveWeightedAvg`; no weighted-average-specific fanout executor should remain.
- [ ] Update comments that describe TSM as having a special dual-query average path.
- [ ] Keep weighted-average arithmetic helpers that remain useful as reducer implementation details.

## Acceptance criteria

- [ ] Every TSM time-series request is represented by a validated plan, including no-aggregation dedup, simple supported aggregations, rank/sort rewrites, weighted `avg`, and unsupported warning fallbacks.
- [ ] No legacy or parallel classification tuple controls request execution outside the plan.
- [ ] Existing successful `avg`, `sum`, `count`, `count_values`, `min`, `max`, dedup, `topk`, `bottomk`, and sort-wrapper tests continue to pass.
- [ ] Weighted `avg` results remain mathematically correct for unequal shard cardinalities in instant and range queries.
- [ ] GET and POST weighted-average rewriting remains equivalent.
- [ ] An `avg` member missing either its sum or count contribution is excluded as a pair and produces an explicit partial-result warning.
- [ ] Variant and member completion order cannot change the reduced result.
- [ ] Existing injected-label stripping, sparse timestamp pairing, unmatched series warnings, headers, status aggregation, cancellation, and capture limits remain covered.
- [ ] `needsDualQuery`, `RewriteForWeightedAvg`, and the dedicated weighted-average dispatch path no longer exist.
- [ ] Adding a future multi-variant operator does not require duplicating the fanout executor or adding an operator-specific dispatch boolean.

## Tests

- [ ] Convert existing classification tests to assert complete plan contents, including variant names, rewritten queries, strategies, reduction, and completeness policy.
- [ ] Retain the current weighted-average tests as characterization tests for successful results, GET/POST rewriting, grouping, wrappers, and sparse timestamps.
- [ ] Add plan validation tests for zero variants, duplicate variant names, missing response authority, invalid strategies, and incompatible reduction configuration.
- [ ] Add generic executor tests for one and multiple variants.
- [ ] Add member-level completeness tests covering missing sum, missing count, parse failure, transport failure, panic, cancellation, and capture-limit failure.
- [ ] Add deterministic-reduction tests with permuted variant completion and member completion order.
- [ ] Add response-authority, winner-header, multi-valued-header, mixed-status, all-failed, and empty-result tests.
- [ ] Run race-enabled tests for the TSM mechanism and Prometheus provider.
- [ ] Benchmark the plan executor against the current standard and weighted-average paths to detect material allocation or latency regressions.

## Code pointers

- [`pkg/backends/tsm_merge_provider.go`](https://github.com/trickstercache/trickster/blob/main/pkg/backends/tsm_merge_provider.go) defines the current classification and weighted-average rewrite contract.
- [`pkg/backends/prometheus/tsm_provider.go`](https://github.com/trickstercache/trickster/blob/main/pkg/backends/prometheus/tsm_provider.go) classifies PromQL queries and rewrites standard-finalizer and weighted-`avg` requests.
- [`pkg/backends/alb/mech/tsm/time_series_merge.go`](https://github.com/trickstercache/trickster/blob/main/pkg/backends/alb/mech/tsm/time_series_merge.go) contains standard fanout, `serveWeightedAvg`, contribution collection, deterministic batch merging, response selection, failure reporting, and finalizer invocation.
- [`pkg/timeseries/dataset/dataset.go`](https://github.com/trickstercache/trickster/blob/main/pkg/timeseries/dataset/dataset.go) contains `FinalizeWeightedAvg` and the dataset merge behavior used by the current reducer.
- [`pkg/backends/alb/mech/tsm/time_series_merge_serve_weighted_avg_test.go`](https://github.com/trickstercache/trickster/blob/main/pkg/backends/alb/mech/tsm/time_series_merge_serve_weighted_avg_test.go) provides weighted-average execution characterization coverage.
- [`pkg/backends/alb/mech/tsm/time_series_merge_weighted_avg_test.go`](https://github.com/trickstercache/trickster/blob/main/pkg/backends/alb/mech/tsm/time_series_merge_weighted_avg_test.go) covers weighted-average finalization and warning behavior.

## Non-goals

- Implementing `stddev`, `stdvar`, `quantile`, `limit_ratio`, or `limitk` in this issue. Those changes will be implemented later in separate Pull Requests.
- Changing successful PromQL aggregation semantics.
- Building a general recursive distributed PromQL planner in this refactor.
- Preserving the old provider interface or weighted-average executor after all in-repository callers and tests have migrated.
- Broad performance redesigns unrelated to making plan execution reusable.

### PR DESCRIPTION
**Pull request by houyuwushang**
## Description

Closes #1073.

This introduces one planning model for every TSM time-series request:

- replace the classification tuple and separate request rewriters with a validated `TSMMergePlan` covering variants, reduction, completeness, response authority, finalization, warnings, and single-member bypass safety
- keep the existing one-variant fanout as a plan-selected fast path, while multi-variant plans use a generic executor that collects by variant and member before accumulation
- migrate weighted `avg` to the generic executor and exclude both sides of a member when either `avg-sum` or `avg-count` is unusable, so reduction never combines different member sets
- share the configured query concurrency limit across variants while retaining independent per-variant fanout capture budgets and metric labels
- preserve response-authority headers/status, partial-hit reporting, cancellation, injected-label stripping, deterministic member order, finalizer ordering, empty responses, and all-authority-failed `502` behavior
- remove the legacy dual-query flag, weighted-average rewrite method, and dedicated weighted-average executor

The new operators from #1069-#1072 are intentionally not included here.

## Type of Change

- - [ ] Bug fix
- - [x] New feature
- - [ ] Optimization
- - [x] Test coverage
- - [ ] Documentation
- - [ ] Infrastructure

## AI Disclosure

- - [ ] This contribution DOES NOT include AI-generated changes
- - [x] This contribution DOES include AI-generated changes, and I have reviewed the relevant contributing guidelines.

## Tests

- affected package tests and race tests for `backends`, Prometheus planning, TSM, and fanout
- `golangci-lint` with the repository configuration
- multi-variant failure coverage for status, parse, transport, panic, cancellation, capture limits, degraded pools, response authority, injected labels, and completion-order determinism
- plan validation coverage and standard/two-variant executor benchmarks

### CODE REVIEW

File: pkg/backends/alb/mech/tsm/merge_plan.go [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+			result.failed = true
+			execution.results[member] = result
+			return
+		}
+		if len(body) > 0 {
+			contribution = prepareGatherContribution(ctx, rsc, body, member, stripKeys)
+		}
+	}
+	result.contrib = contribution
+	result.failed = contribution == nil
```
Comment (jranson) [inline pkg/backends/alb/mech/tsm/merge_plan.go]: because prometheus returns a valid response envelope even on 5xx errors (e.g, a 503), currently error status responses can potentially decode to a non-nil dataset and be included in the reduction, which could impact weighted averages and other operations. To protect against that, here you can check for a valid status code (2xx only) and consider the result failed if the code is not 2xx:

```
result.failed = contribution == nil || !result.statusCode >= 200 && result.statusCode < 300
```

If the result failed but has a non-nil DataSet, you can keep the decoded contribution temporarily so it can supply an error-envelope `responseSeed` and  `applyPlanCompleteness` will remove it from reduction.

It would be good to add a regression test using a non-nil error DataSet or Prometheus error envelope to verify this.

File: pkg/backends/tsm_merge_provider.go [hunk 2]
```diff
@@ -16,32 +16,184 @@
...
+			return fmt.Errorf("tsm merge plan variant %q has no request", variant.Name)
+		}
+		if _, ok := requestPointers[variant.Request]; ok {
+			return fmt.Errorf("tsm merge plan variants share request %q", variant.Name)
+		}
+		requestPointers[variant.Request] = struct{}{}
+		// dataset.MergeStrategy currently defines values 0 through 5. Keep the
+		// shallow backends package independent of dataset while rejecting values
+		// that the merge layer would otherwise silently treat as deduplication.
+		if variant.MergeStrategy < 0 || variant.MergeStrategy > 5 {
```
Comment (jranson) [inline pkg/backends/tsm_merge_provider.go]: It would be good to define in dataset/merge_strategy.go a const for `MaxMergeStrategyValue = MergeStrategyCount`, and use `dataset.MaxMergeStrategyValue` instead of a literal 5 here. that way if we add more merge strategies later, we can update the MaxMergeStrategyValue const assignment to the new largest MergeStrategy* const value without having to modify the actual validation logic.

File: pkg/backends/tsm_merge_provider_test.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+			},
+			{
+				Name:          "avg-count",
+				Request:       testTSMRequest("count"),
+				MergeStrategy: 1,
+			},
+		},
+		Reduction: TSMReductionSpec{
+			Kind:          TSMReductionWeightedAverage,
+			InputVariants: []string{"avg-sum", "avg-count"},
```
Comment (jranson) [inline pkg/backends/tsm_merge_provider_test.go]: While this works because `reducePlan` interprets the positions implicitly (e.g., `countAccumulator := accumulators[indexes[plan.Reduction.InputVariants[1]]]`), a plan like `InputVariants: []string{"avg-count", "avg-sum"}` (where the variants were inadvertently reversed) would currently pass validation even though the ordering is incorrect.

At minimum, we should document the positional contract in TSMReductionSpec and validate that weighted-average inputs use the expected variant strategies in the correct order. But the InputVariants slice will become even more brittle when we need to use three-input variance reductions for forthcoming reductions like stddev. One way to ensure the InputVariants positions are properly configured would be to add and use helper functions (e.g.,  `func TSMReductionWeightedAverageVariants() []string { return []string{"avg-sum", "avg-count"} }`)  anywhere the InputVariants slices need to be defined (in both production code and tests).  Also, it might be a good idea to define and use consts for the variant names like `avg-sum` instead of using the literals throughout the codebase, to avoid typos and other human error.

**Review by jranson — CHANGES_REQUESTED** [review]
@houyuwushang  this refactor is solid and largely meets the requested criteria. just a few minor things to fix up.

File: pkg/backends/tsm_merge_provider_test.go (hunk 3 above)
Comment (houyuwushang) [reply pkg/backends/tsm_merge_provider_test.go]: Added weighted-average variant constants and TSMReductionWeightedAverageVariants(), and used them in provider and test plan construction. Validation now rejects reversed inputs and requires both variants to use MergeStrategySum.

File: pkg/backends/tsm_merge_provider.go (hunk 2 above)
Comment (houyuwushang) [reply pkg/backends/tsm_merge_provider.go]: Added dataset.MaxMergeStrategyValue and switched plan validation to use it. The new dependency exposed a test-only cycle through the SQL test package importing the broad testutil package, so I localized the two fixed time values used there.

File: pkg/backends/alb/mech/tsm/merge_plan.go (hunk 1 above)
Comment (houyuwushang) [reply pkg/backends/alb/mech/tsm/merge_plan.go]: Handled in 9d2d04db. Non-2xx results are now marked failed after decoding, so the DataSet remains available for response seeding while completeness removes that member from every reduction input. I added coverage for both a mixed pool and a no-complete-member 503 response.

File: pkg/backends/prometheus/tsm_provider.go [hunk 4]
```diff
@@ -33,14 +34,112 @@ import (
...
+		switch agg {
+		case "sum", "count", "count_values":
+			strategy = int(dataset.MergeStrategySum)
+		case "avg":
+			return weightedAveragePlan(r, query, fanoutQuery, finalizer)
+		case "min":
+			strategy = int(dataset.MergeStrategyMin)
+		case "max":
+			strategy = int(dataset.MergeStrategyMax)
+		case "stddev", "stdvar", "quantile", "topk", "bottomk", "limitk", "limit_ratio":
```
Comment (jranson) [inline pkg/backends/prometheus/tsm_provider.go:66]: should `topk` and `bottomk` be included in the unsupported case?

Comment (jranson) [reply pkg/backends/prometheus/tsm_provider.go:66]: it looks like these are just for the inner aggregators, so the warning is appropriate.

**Review by jranson — APPROVED** [review]
